### PR TITLE
Add an updated proposal for DOM Parts.

### PR DESCRIPTION
**Pull request by rniwa**

### CODE REVIEW

**Comment by rniwa** [comment]
@annevk @justinfagnani @hober @mfreed7

**Comment by justinfagnani** [comment]
cc @yuzhe-han

**Comment by rniwa** [comment]
@justinfagnani @yuzhe-han : would you like me to add acknowledgment for your involvement? I wasn't sure if you'd like to put your names on it or not.

**Comment by yuzhe-han** [comment]
@rniwa : sure. I enjoyed our discussions and look forward to collaborating in the future.   Thanks.

**Comment by justinfagnani** [comment]
That would be great, thanks!

File: proposals/DOM-Parts.md [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+```
+
+After running this code, `instance` would contain a DOM tree equivalent of:
+```html
+<section><h1 id="name"></h1>Email: <a id="link" href=""></a></section>
+```
+
+Then we can assign values to various parts we’ve just created as follows:
+```js
+updateParts(parts, {name: "Ryosuke Niwa", email: "rniwa@webkit.org", "mailAddress": "mailto:rniwa@webkit.org"})
```
Comment (stramel) [inline proposals/DOM-Parts.md]:
```suggestion
updateParts(parts, {name: "Ryosuke Niwa", email: "rniwa@webkit.org", "emailAddress": "mailto:rniwa@webkit.org"})
```

File: proposals/DOM-Parts.md [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+interface Part {
+    attribute any value;
+    void commit();
+};
+
+interface NodePart : Part {
+    readonly attribute Node node;
+};
+
+interface AttributePart : NodePart {
```
Comment (annevk) [inline proposals/DOM-Parts.md]: It's a bit confusing to me that an attribute is a node (whereas the DOM mostly tries to pretend it's not) and a child node is not. Perhaps that's not what NodePart means, but you would think it is given how AttributePart and ChildNodePart are named.

File: proposals/DOM-Parts.md [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+Then assigning values as follows will update the DOM:
+namePart.value = "Ryosuke Niwa"
+emailPart.value = "rniwa@webkit.org"
+emailAttributePart.value = "mailto:rniwa@webkit.org"
+```
+
+The resultant DOM will look like this:
+```html
+<section><h1 id="name">Ryosuke Niwa</h1>Email: <a id="link" href="mailto:rniwa@webkit.org"></a></section>
+```
```
Comment (annevk) [inline proposals/DOM-Parts.md:120]: Why does this work without commit?

**Review by annevk — COMMENTED** [review]
I still need to read more, but this looks interesting and seems like a good iteration given the previous feedback of the approach being too high-level. Once thing that's not entirely clear to me is what the internal model of a part is and why having it is beneficial other than developer ergonomics. Does this enable any performance wins over mutating the tree yourself, for instance?

File: proposals/DOM-Parts.md (hunk 1 above)
Comment (rniwa) [reply proposals/DOM-Parts.md]: Will fix.

File: proposals/DOM-Parts.md (hunk 2 above)
Comment (rniwa) [reply proposals/DOM-Parts.md]: Oh oops, that's a typo. It should inherit from `Part`

File: proposals/DOM-Parts.md (hunk 3 above)
Comment (rniwa) [reply proposals/DOM-Parts.md:120]: Oh oops, that's because I forgot to add calls to `commit()`. Will fix.

**Comment by rniwa** [comment]
> I still need to read more, but this looks interesting and seems like a good iteration given the previous feedback of the approach being too high-level. Once thing that's not entirely clear to me is what the internal model of a part is and why having it is beneficial other than developer ergonomics. Does this enable any performance wins over mutating the tree yourself, for instance?

Performance improvement is possible if we introduce some kind of batching.

**Comment by rniwa** [comment]
@justinfagnani @hober @mfreed7 @yuzhe-han : Added the acknowledgement.  Let's merge this so that people can take a look before tomorrow's meeting.

**Comment by justinfagnani** [comment]
> Once thing that's not entirely clear to me is what the internal model of a part is and why having it is beneficial other than developer ergonomics. Does this enable any performance wins over mutating the tree yourself, for instance?

Another big win for template libraries and hand-written DOM code is the cloning behavior. This would be the first way to clone some kind of _reference_ to nodes, along with a subtree.

**Comment by rniwa** [comment]
Can someone approve this PR?

**Comment by leobalter** [comment]
I'm interested but I'm trying to fit in time this afternoon to read it.

File: proposals/DOM-Parts.md [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+
+* **Pros**: Simplicity.
+* **Cons**: Coming up with a nice syntax to create a sequence of `ChildNodePart` and string can be tricky.
+
+### Option 2. Introduce `ChildNodePartGroup`
+
+In this approach, we group multiple `ChildNodePart`s together by creating an explicit group:
+
+```
```
Comment (myakura) [inline proposals/DOM-Parts.md]:
```suggestion
```js
```

File: proposals/DOM-Parts.md [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+<section><h1 id="name"></h1>Email: <a id="link" href=""></a></section>
+```
+
+Then we can create `ChildNodePart` for [`h1` element](https://html.spec.whatwg.org/multipage/sections.html#the-h1,-h2,-h3,-h4,-h5,-and-h6-elements)
+and [`a` element](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element)
+and `AttributePart` for [`a` element](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element) as follows: 
+```js
+namePart = new ChildNodePart(name)
+emailPart = new ChildNodePart(link)
+emailAttributePart = new AttributePart(link, ’href’)
```
Comment (myakura) [inline proposals/DOM-Parts.md]: it looks single quotes are unintentionally converted to smart quotes throughout the draft.

**Comment by annevk** [comment]
If people want to read it they can do so at https://github.com/w3c/webcomponents/blob/405d041d8346355d11c1f1d426407488e7c16820/proposals/DOM-Parts.md. It does seem good to address the issues @myakura raised before merging. What's the best way to raise high-level issues? File a new issue with the templates label?

File: proposals/DOM-Parts.md (hunk 4 above)
Comment (rniwa) [reply proposals/DOM-Parts.md]: Will fix.

**Comment by rniwa** [comment]
> It does seem good to address the issues @myakura raised before merging.

Fixed. Can we merge it now? It would be really unhelpful for this PR to be continue to be open while more people keep adding more review comments here. It would be far better if this was merged into the repository and we filed issues instead.

**Comment by rniwa** [comment]
> What's the best way to raise high-level issues? File a new issue with the templates label?

File issues with [DOM Parts] probably since this new proposal is no longer anything to do with template elements per se.

**Comment by rniwa** [comment]
Alright, I added this to [wiki page](https://github.com/w3c/webcomponents/wiki/DOM-Part-API---First-Step-of-Template-Instantiation) instead since this isn't getting approved by anyone.

**Comment by annevk** [comment]
You gave me 8 minutes to reply? I'm happy to merge this in a bit.

**Review by annevk — APPROVED** [review]

**Comment by rniwa** [comment]
> You gave me 8 minutes to reply? I'm happy to merge this in a bit.

Oh sorry, I thought you went off somewhere. Thanks for merging it!

**Pull request merged**